File: scripts/zigux/check-phase1-find-bit-bench-uniqueness.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
from pathlib import Path
import tempfile


HERE = Path(__file__).resolve()
DEFAULT_ROOT = HERE.parents[2] if len(HERE.parents) > 2 else HERE.parent
PHASE1_BENCH_REL = Path("zigux/tests/phase1_bench.zig")

FIND_BIT_BENCH_MARKERS = {
    "find_bit_bench_fn": "fn findBitBench() struct { checksum: u64 } {",
    "find_bit_edge_fn": "fn findBitEdgeBench() struct { checksum: u64 } {",
    "find_bit_bench_call": "const find_bit_result = findBitBench();",
    "find_bit_edge_call": "const find_bit_edge_result = findBitEdgeBench();",
    "find_next_iterations_print": 'try stdout_writer.interface.print("PHASE1_BENCH_FIND_NEXT_BIT_ITERATIONS={d}\\n", .{iterations_find_bit});',
    "find_next_checksum_print": 'try stdout_writer.interface.print("PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM={d}\\n", .{find_bit_result.checksum});',
    "find_edge_iterations_print": 'try stdout_writer.interface.print("PHASE1_BENCH_FIND_BIT_EDGE_ITERATIONS={d}\\n", .{iterations_find_bit_edge});',
    "find_edge_checksum_print": 'try stdout_writer.interface.print("PHASE1_BENCH_FIND_BIT_EDGE_CHECKSUM={d}\\n", .{find_bit_edge_result.checksum});',
    "boundary_next_bit": "checksum +%= @intCast(find_bit.findNextBit(&boundary_set, head_nbits, boundary));",
    "boundary_next_and_bit": "checksum +%= @intCast(find_bit.findNextAndBit(&boundary_set, &boundary_set, head_nbits, boundary));",
    "boundary_next_zero_bit": "checksum +%= @intCast(find_bit.findNextZeroBit(&boundary_zero, head_nbits, boundary));",
    "tail_first_bit": "checksum +%= @intCast(find_bit.findFirstBit(&tail_set, tail_nbits));",
    "tail_first_and_bit": "checksum +%= @intCast(find_bit.findFirstAndBit(&tail_set, &tail_set, tail_nbits));",
    "tail_last_bit": "checksum +%= @intCast(find_bit.findLastBit(&tail_set, tail_nbits));",
}


def repo_root(root: str | None) -> Path:
    return Path(root).resolve() if root else DEFAULT_ROOT.resolve()


def bench_source_path(root: Path) -> Path:
    return root / PHASE1_BENCH_REL


def validate_bench_source(text: str) -> tuple[str, list[str]]:
    missing: list[str] = []
    duplicate: list[str] = []
    for label, marker in FIND_BIT_BENCH_MARKERS.items():
        count = text.count(marker)
        if count == 0:
            missing.append(label)
        elif count > 1:
            duplicate.append(f"{label}:count={count}")
    if missing:
        return ("missing_find_bit_bench_markers", missing)
    if duplicate:
        return ("duplicate_find_bit_bench_markers", duplicate)
    return ("pass", [])


def load_runtime_bench_source(path: Path) -> tuple[str, object]:
    try:
        text = path.read_text(encoding="utf-8")
    except FileNotFoundError:
        return ("missing_bench_source_file", path)
    kind, payload = validate_bench_source(text)
    return (kind, payload)


def build_sample_source(omit_label: str | None = None, duplicate_label: str | None = None) -> str:
    lines = [
        "fn findBitBench() struct { checksum: u64 } {",
        "    checksum +%= @intCast(find_bit.findNextBit(&boundary_set, head_nbits, boundary));",
        "    checksum +%= @intCast(find_bit.findNextAndBit(&boundary_set, &boundary_set, head_nbits, boundary));",
        "    checksum +%= @intCast(find_bit.findNextZeroBit(&boundary_zero, head_nbits, boundary));",
        "}",
        "fn findBitEdgeBench() struct { checksum: u64 } {",
        "    checksum +%= @intCast(find_bit.findFirstBit(&tail_set, tail_nbits));",
        "    checksum +%= @intCast(find_bit.findFirstAndBit(&tail_set, &tail_set, tail_nbits));",
        "    checksum +%= @intCast(find_bit.findLastBit(&tail_set, tail_nbits));",
        "}",
        "const find_bit_result = findBitBench();",
        "const find_bit_edge_result = findBitEdgeBench();",
        'try stdout_writer.interface.print("PHASE1_BENCH_FIND_NEXT_BIT_ITERATIONS={d}\\n", .{iterations_find_bit});',
        'try stdout_writer.interface.print("PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM={d}\\n", .{find_bit_result.checksum});',
        'try stdout_writer.interface.print("PHASE1_BENCH_FIND_BIT_EDGE_ITERATIONS={d}\\n", .{iterations_find_bit_edge});',
        'try stdout_writer.interface.print("PHASE1_BENCH_FIND_BIT_EDGE_CHECKSUM={d}\\n", .{find_bit_edge_result.checksum});',
    ]

    if omit_label is not None:
        marker = FIND_BIT_BENCH_MARKERS[omit_label]
        lines = [line for line in lines if marker not in line]
    if duplicate_label is not None:
        marker = FIND_BIT_BENCH_MARKERS[duplicate_label]
        for idx, line in enumerate(lines):
            if marker in line:
                lines.insert(idx + 1, line)
                break
    return "\n".join(lines) + "\n"


def assert_case(condition: bool, name: str, payload: object = None) -> None:
    if not condition:
        raise AssertionError((name, payload))


def run_self_test() -> None:
    case_count = 0

    kind, payload = validate_bench_source(build_sample_source())
    assert_case(kind == "pass", "baseline", (kind, payload))
    case_count += 1

    for label in FIND_BIT_BENCH_MARKERS:
        kind, payload = validate_bench_source(build_sample_source(omit_label=label))
        assert_case(kind == "missing_find_bit_bench_markers", f"missing {label}", (kind, payload))
        assert_case(payload == [label], f"missing payload {label}", payload)
        case_count += 1

    for label in FIND_BIT_BENCH_MARKERS:
        kind, payload = validate_bench_source(build_sample_source(duplicate_label=label))
        assert_case(kind == "duplicate_find_bit_bench_markers", f"duplicate {label}", (kind, payload))
        assert_case(payload == [f"{label}:count=2"], f"duplicate payload {label}", payload)
        case_count += 1

    with tempfile.TemporaryDirectory(prefix="phase1-find-bit-bench-unique-") as tmp:
        source_path = Path(tmp) / PHASE1_BENCH_REL
        kind, payload = load_runtime_bench_source(source_path)
        assert_case(kind == "missing_bench_source_file", "missing runtime source", (kind, payload))
        case_count += 1

        source_path.parent.mkdir(parents=True, exist_ok=True)
        source_path.write_text(build_sample_source(), encoding="utf-8")
        kind, payload = load_runtime_bench_source(source_path)
        assert_case(kind == "pass", "runtime source pass", (kind, payload))
        case_count += 1

    print("PHASE1_FIND_BIT_BENCH_UNIQUENESS_SELF_TEST=pass")
    print(f"PHASE1_FIND_BIT_BENCH_UNIQUENESS_SELF_TEST_CASE_COUNT={case_count}")


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Validate that the Phase 1 find_bit bench source keeps each required marker exactly once."
    )
    parser.add_argument("--repo-root", "--root", dest="repo_root", help="Override the repository root used for validation.")
    parser.add_argument("--self-test", action="store_true", help="Run checker self-test cases only.")
    args = parser.parse_args()

    if args.self_test:
        run_self_test()
        return 0

    root = repo_root(args.repo_root)
    phase1_bench = bench_source_path(root)
    kind, payload = load_runtime_bench_source(phase1_bench)
    if kind != "pass":
        print("PHASE1_FIND_BIT_BENCH_UNIQUENESS=fail")
        print(f"PHASE1_FIND_BIT_BENCH_UNIQUENESS_REASON={kind}")
        print(payload)
        return 1

    print("PHASE1_FIND_BIT_BENCH_UNIQUENESS=pass")
    print(f"PHASE1_FIND_BIT_BENCH_UNIQUENESS_SOURCE={phase1_bench}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase1_find_bit_bench_uniqueness.zig
// Ported from check-phase1-find-bit-bench-uniqueness.py by port_phase1_guards.py
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const pass_marker = "PHASE1_FIND_BIT_BENCH_UNIQUENESS_SELF_TEST=pass";

const FIND_BIT_BENCH_MARKERS = [_]struct { label: []const u8, marker: []const u8 }{
    .{ .label = "find_bit_bench_fn", .marker = "fn findBitBench() struct { checksum: u64 } {" },
    .{ .label = "find_bit_edge_fn", .marker = "fn findBitEdgeBench() struct { checksum: u64 } {" },
    .{ .label = "find_bit_bench_call", .marker = "const find_bit_result = findBitBench();" },
    .{ .label = "find_bit_edge_call", .marker = "const find_bit_edge_result = findBitEdgeBench();" },
    .{ .label = "find_next_iterations_print", .marker = "try stdout_writer.interface.print(\"PHASE1_BENCH_FIND_NEXT_BIT_ITERATIONS={d}\\n\", .{iterations_find_bit});" },
    .{ .label = "find_next_checksum_print", .marker = "try stdout_writer.interface.print(\"PHASE1_BENCH_FIND_NEXT_BIT_CHECKSUM={d}\\n\", .{find_bit_result.checksum});" },
    .{ .label = "find_edge_iterations_print", .marker = "try stdout_writer.interface.print(\"PHASE1_BENCH_FIND_BIT_EDGE_ITERATIONS={d}\\n\", .{iterations_find_bit_edge});" },
    .{ .label = "find_edge_checksum_print", .marker = "try stdout_writer.interface.print(\"PHASE1_BENCH_FIND_BIT_EDGE_CHECKSUM={d}\\n\", .{find_bit_edge_result.checksum});" },
    .{ .label = "boundary_next_bit", .marker = "checksum +%= @intCast(find_bit.findNextBit(&boundary_set, head_nbits, boundary));" },
    .{ .label = "boundary_next_and_bit", .marker = "checksum +%= @intCast(find_bit.findNextAndBit(&boundary_set, &boundary_set, head_nbits, boundary));" },
    .{ .label = "boundary_next_zero_bit", .marker = "checksum +%= @intCast(find_bit.findNextZeroBit(&boundary_zero, head_nbits, boundary));" },
    .{ .label = "tail_first_bit", .marker = "checksum +%= @intCast(find_bit.findFirstBit(&tail_set, tail_nbits));" },
    .{ .label = "tail_first_and_bit", .marker = "checksum +%= @intCast(find_bit.findFirstAndBit(&tail_set, &tail_set, tail_nbits));" },
    .{ .label = "tail_last_bit", .marker = "checksum +%= @intCast(find_bit.findLastBit(&tail_set, tail_nbits));" },
};

const PHASE1_BENCH_REL = "zigux/tests/phase1_bench.zig";

fn collectFailures(
    io: Io,
    allocator: std.mem.Allocator,
    root: []const u8,
) !std.ArrayList([]const u8) {
    var failures: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    {
        const relative_path = "zigux/tests/phase1_bench.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    if (failures.items.len > 0) return failures;

    {
        const relative_path = "zigux/tests/phase1_bench.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        const text = guard.readUtf8File(io, allocator, full_path) catch |err| switch (err) {
            guard.GuardError.IOError => {
                const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
                try failures.append(allocator, issue);
                return failures;
            },
            else => return err,
        };
        defer allocator.free(text);
        for (FIND_BIT_BENCH_MARKERS) |entry| {
            const count = guard.countOccurrences(text, entry.marker);
            if (count != 1) {
                const issue = try std.fmt.allocPrint(allocator, "{s}:expected=1:actual={d}", .{ entry.label, count });
                try failures.append(allocator, issue);
            }
        }
    }

    return failures;
}

fn buildSampleSource(allocator: std.mem.Allocator) ![]u8 {
    var content = std.ArrayList(u8).empty;
    errdefer content.deinit(allocator);
    for (FIND_BIT_BENCH_MARKERS) |entry| {
        try content.appendSlice(allocator, entry.marker);
        try content.append(allocator, '\n');
    }
    return try content.toOwnedSlice(allocator);
}

pub fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    var tmp = try guard.TempWorkspace.init(io, allocator, "selftest");
    defer tmp.deinit();
    const root = try tmp.rootPath(allocator);
    defer allocator.free(root);
    const sample = try buildSampleSource(allocator);
    defer allocator.free(sample);
    {
        const relative_path = "zigux/tests/phase1_bench.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        try guard.writeUtf8File(io, full_path, sample);
        var failures = try collectFailures(io, allocator, root);
        defer {
            for (failures.items) |item| allocator.free(item);
            failures.deinit(allocator);
        }
        try guard.expectSelfTest(failures.items.len == 0);
    }
    try guard.printLine(io, "{s}", .{pass_marker});
    try guard.printLine(io, "PHASE1_FIND_BIT_BENCH_UNIQUENESS_SELF_TEST_CASE_COUNT={d}", .{@as(usize, 31)});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var explicit_root: ?[]const u8 = null;
    var self_test = false;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    const root = if (explicit_root) |value| value else try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);

    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    if (failures.items.len > 0) {
        try guard.printLine(io, "PHASE1_FIND_BIT_BENCH_UNIQUENESS_SELF_TEST=fail", .{});
        for (failures.items) |failure| try guard.printLine(io, "{s}", .{failure});
        std.process.exit(1);
    }

    try guard.printLine(io, "{s}", .{pass_marker});
    std.process.exit(0);
}
